Add Best Practice: Don't Assume Plugin Order (#34311)
gradle/gradle: platforms/documentation/docs/src/docs/userguide/reference/best-practices/best_practices_general.adoc

`<<tags_reference.adoc#tag:structuring-builds,#structuring-builds>>`

[[dont_assume_plugin_order]]
== Don’t Assume your Plugin is Applied after Another

> Gradle does not guarantee the order in which plugins are applied, especially across multiple build scripts, projects, convention plugins, or included builds.

Reviewer: Gradle does guarantee order, but that order might not be obvious from looking at a single place.

As a result, you should not write build logic or plugins that depend on a specific plugin application order.

=== Explanation

In a single `build.gradle(.kts)` file, plugins appear to be applied sequentially:

[source,kotlin]
.build.gradle.kts
----
plugins {
    id("pluginA")
    id("pluginB")
}
----

> However, in <<multi_project_builds.adoc#multi_project_builds,multi-project builds>>, with multiple `build.gradle(.kts)` files or a <<plugins.adoc#sec:convention_plugins,convention plugin>>, plugin application can be non-deterministic.

Reviewer: > plugin application can be non-deterministic.

order is deterministic, but not always obvious because plugins can apply other plugins.

> However, in <<multi_project_builds.adoc#multi_project_builds,multi-project builds>>, with multiple `build.gradle(.kts)` files or a <<plugins.adoc#sec:convention_plugins,convention plugin>>, plugin application can be non-deterministic.

Author: Already updated based on @tresat's feedback. LMK what you think.

Gradle may configure projects in parallel, and the effective order can vary between runs:

[source,kotlin]

> .convention-build.gradle.kts

Author: I thought it was a good idea. I renamed it to `project`.

----
plugins {
    id("pluginC")
    id("pluginD")
}
----

> You cannot assume whether `pluginA`, `pluginB`, `pluginC`, or `pluginD` will be applied first.

Author: I will update it to say "what if PluginA applied pluginD?"

==== Build Engineers

Writing build logic that assumes plugin ordering can lead to:

- Fragile builds that break when project structure changes.

> - Brittle behavior that works locally but fails in CI or in <<composite_builds.adoc#composite_builds,composite builds>>.

Reviewer: Plugin behaving differently in CI vs local sounds like a whole other kind of problem.

I think we should just drop it.

> - Hard-to-diagnose bugs when plugins interact inconsistently due to race conditions.

Author: Added.

Techniques like `allprojects {}`, `subprojects {}`, or `afterEvaluate {}` do not reliably control plugin order across projects or included builds.

Avoid build logic that assumes ordering between different projects, included builds, or applied scripts, the order is not guaranteed.

==== Plugin Developers

If your plugin requires another plugin, do not assume it has already been applied.
Either:

- Apply it yourself, or
- Fail fast with a clear error if it is missing.

[source,groovy]
----
if (!project.pluginManager.hasPlugin(requiredPlugin)) {
    // Warn or fail with a clear message
}
project.pluginManager.apply('com.example.required-plugin')
----

> If your plugin can work with or without another plugin, and behavior changes when that plugin is present, use link:{javadocPath}/org/gradle/api/plugins/PluginManager.html#withPlugin(java.lang.String,org.gradle.api.Action)[`withPlugin()`]: to defer configuration until it is actually applied:

Author: Added.

`<<tags_reference.adoc#tag:structuring-builds,#structuring-builds>>`

[[dont_assume_plugin_order]]
== Don’t Assume your Plugin is Applied after Another

> Gradle's plugin application is deterministic, but the effective execution order is complex and not guaranteed to follow a visible sequence, especially across multiple build scripts, projects, convention plugins, or included builds.

Reviewer 2: ```suggestion
Gradle's plugin application is deterministic but opaque.
It is difficult to reason about, especially across multiple build scripts, projects, convention plugins, or included builds.
```

As a result, you should not write build logic or plugins that depend on a specific plugin application order.

=== Explanation

In a single `build.gradle(.kts)` file, plugins appear to be applied sequentially:

[source,kotlin]
.build.gradle.kts
----
plugins {
    id("pluginA")
    id("pluginB")
}
----

However, in <<multi_project_builds.adoc#multi_project_builds,multi-project builds>>, with multiple `build.gradle(.kts)` files or a <<plugins.adoc#sec:convention_plugins,convention plugin>>, plugin application can be hard to determine:

[source,kotlin]

> .project.gradle.kts

Reviewer: What is this project.gradle.kts? 

How does it apply for the single build.gradle.kts shown above?

Reviewer 2: I'm also not sure how this is meant to relate to the build file above.  This needs more clarity.

Author: more copy pasta. fixed.

----
plugins {
    id("pluginC")
    id("pluginD")
}
----

> You cannot assume whether `pluginA`, `pluginB`, `pluginC`, or `pluginD` will be applied first because the first thing `pluginA` does could be to apply `pluginD`.

Reviewer: > `pluginA` does could be to apply `pluginD`.

`pluginA` could  apply `pluginD`.

====
[.multi-language-sample]
=====
[source, kotlin]
----
// Configure behavior that depends on required-plugin
project.pluginManager.withPlugin("com.example.required-plugin") {  }

project.plugins.configureEach { plugin ->
    when (plugin) {
        // Configure behavior specific to RequiredPlugin
        is com.example.RequiredPlugin -> {  }
        // Configure behavior specific to OtherPlugin
        is com.example.OtherPlugin -> {  }
    }
}
----
=====
[.multi-language-sample]
=====
[source, groovy]
----
// Configure behavior that depends on required-plugin
project.pluginManager.withPlugin('com.example.required-plugin') {  }

project.plugins.configureEach { plugin ->
    // Configure behavior specific to RequiredPlugin
    if (plugin instanceof com.example.RequiredPlugin) {  }
    // Configure behavior specific to OtherPlugin
    else if (plugin instanceof com.example.OtherPlugin) {  }

> }

Reviewer 2: ❗ I don't think we should include `configureEach`.  `withPlugin` is the preferred method in almost every case.  If we want to help people choose between alternatives we should firmly recommend that.

Reviewer 2: I forgot about that case.  I think we should make it clear that if an id is present, you should use `withPlugin`.  This provides clear guidance of which alternative to choose most of the time.

====
include::sample[dir="snippets/bestPractices/dontAssumePluginOrder-avoid/kotlin",files="app/build.gradle.kts[tags=avoid-this]"]
====

In the plugin implementation, `MyPlugin.kt` also assumes `java` is already applied:

====
include::sample[dir="snippets/bestPractices/dontAssumePluginOrder-avoid/kotlin",files="buildSrc/src/main/kotlin/MyPlugin.kt[tags=avoid-this]"]
====

==== Do This Instead

The fixed version removes `afterEvaluate` and avoids assumptions about when or where the `java` plugin is applied.

The root build file has been deleted as there is no longer a need to use `subprojects {}`.

In the `app` subproject, the build file uses `plugins.withId("java") {}` to safely configure tasks once `java` is applied.

====
include::sample[dir="snippets/bestPractices/dontAssumePluginOrder-do/kotlin",files="app/build.gradle.kts[tags=do-this]"]
====

In the plugin implementation, `MyPlugin.kt` explicitly applies the `java` plugin:

====
include::sample[dir="snippets/bestPractices/dontAssumePluginOrder-do/kotlin",files="buildSrc/src/main/kotlin/MyPlugin.kt[tags=do-this]"]
====

=== References

- <<plugins_intermediate.adoc#sec:using_plugins,Using Plugins>>

=== Tags

> `<<tags_reference.adoc#tag:structuring-builds,#structuring-builds>>`

Reviewer 2: Also `#plugins`?
